### PR TITLE
Arm reworking

### PR DESCRIPTION
**Pull request by anton-bondarev**

### CODE REVIEW

File: src/arch/arm/armlib/entry.S [hunk 1]
```diff
@@ -8,60 +8,80 @@
...
+	/* save temp register*/
+	stmfd  sp!, {r0}
+	/* storing LP */
+	stmfd  sp!, {lr}
+	/*storing SPSR and valuable regs*/
+	mrs    lr, SPSR
+	/* storing SPSR */
+	stmfd  sp!, {lr}
+	/* pointer to saved on special stack registers */
+	mov    r0, sp
```
Comment (AntonKozlov) [inline src/arch/arm/armlib/entry.S]: After this stack is 
```

[r0]
[lr]
[SPSR] <- r0 points here
```
...

File: src/arch/arm/armlib/entry.S [hunk 2]
```diff
@@ -8,60 +8,80 @@
...
-	ldr	r0, =this_spsr_lr
-	ldmia	r0, {r1, r2}
-	stmfd	sp!,{r0, r1, r2}
+	/* now we have previous mode and sp */
+	stmfd  sp!, {r1-r12, lr}
 
-	ldr	lr, =irq_leave
-	msr	CPSR, #ARM_MODE_IRQ | I_BIT | F_BIT
+	/* restore spsr, lr, tmp reg */
+	ldmfd  r0!, {r1} /* spsr */
```
Comment (AntonKozlov) [inline src/arch/arm/armlib/entry.S]: This starts to load in decreasing addressed. After load stack will be 
```
[r0]
[lr]
[SPSR] 
... <- r0 points here
```

I suppose `ldmia r0, {r1, r2, r3}` should be used.

Comment (AntonKozlov) [inline src/arch/arm/armlib/entry.S]: Ops, ldmfd is ldmia. So this line is correct

File: src/arch/arm/armlib/entry.S (hunk 1 above)
Comment (anton-bondarev) [inline src/arch/arm/armlib/entry.S]: I think this code is correct.
Why can 'r0' be changed?

**Pull request merged**